userspace/libsinsp/parsers.cpp

```diff
@@ -4522,38 +4536,61 @@ void sinsp_parser::parse_setpgid_exit(sinsp_evt *evt)
 		pgid = *(int64_t *)parinfo->m_val;
 		ASSERT(parinfo->m_len == sizeof(int64_t));
 
-		//
-		// If pid is zero, then the process ID of the calling process is used.
-		//
-		sinsp_threadinfo* ptinfo = NULL;
-		if (pid == 0)
+		// Determine whether the caller is in a pid namespace
+		// or not. That determines how we should interpret the
+		// pid arg.
+		if(cinfo->m_pid == cinfo->m_vpid)
 		{
-			ptinfo = evt->get_thread_info();
+			// The caller is not in a pid namespace, so
+			// the pid is already in the global pid
+			// namespace.
+			gpid = pid;
 		}
 		else
 		{
-			ptinfo = m_inspector->get_thread(pid, false, true);
+			// pidarg is relative to the namespace of the
+			// caller. Convert it to the global namespace.
+			gpid = (pid == 0 ?
+				0 :
+				m_inspector->m_thread_manager->get_pid_for_vpid(cinfo->m_container_id, pid));
+
+			if(gpid == -1)
+			{
+				ASSERT(false);
+				return;
+			}
+		}
+
+		//
+		// If pid is zero, then the process ID of the calling process is used.
+		//
+		if (gpid == 0)
+		{
+			gpid = cinfo->m_pid;
 		}
 
-		if(ptinfo == NULL)
+		// Now look up the target threadinfo
+		tinfo = m_inspector->get_thread(gpid, false, true);
+
+		if(!tinfo)
 		{
 			ASSERT(false);
 			return;
 		}
 
 		//
-		// If pgid is zero, then the PGID of the process specified
-		// by pid is made the same as its process ID.
+		// If pgid is zero, then the PGID of the process
+		// specified by pid is made the same as its process ID
+		// (in its pid namespace).
 		//
 		if (pgid == 0)
 		{
-			ptinfo->m_pgid = ptinfo->m_pid;
-		}
-		else
-		{
-			ptinfo->m_pgid = pgid;
+			pgid = tinfo->m_vpid;
 		}
 
+		// Now actually change the target's vpgid to that
+		// of pinfo.
+		tinfo->m_vpgid = pgid;
```

Comment:
I'm wondering if we should do some translation on the `pgid` value? I'm not sure if it's really a use case, but in the documentation of `proc.vpgid` you say `... as seen from its current PID namespace ...`, however that is not really true when you deal with a child in a different pid namespace. For example:

```
#define _GNU_SOURCE

#include <stdio.h>
#include <sys/types.h>
#include <unistd.h>
#include <sched.h>

int main(int argc, char *argv[])
{
        unshare(CLONE_NEWPID);

        int pid = fork();

        if(pid)
                setpgid(pid, pid));

        while(1)
                sleep(1);

        return 0;
}
```

If this program is launched before sysdig, then the setpgid system call will have as `pgid` argument the pid of the child as seen in the global pid namespace, so its value will be set to that. If instead this program is launched before sysdig, the `pgid` set for the child thread will be its current virtual pid, since it comes from scap.